api/main.go

```diff
@@ -152,3 +153,195 @@ func (d *Dispatcher) Write(ctx context.Context, vl *ValueList) error {
+		}
+		for i := range c.Children {
+			// If a config has children but the struct has no corresponding field, or the corresponding field is an
+			// unexported struct field the child is ignored without notice.
+			if field := drv.FieldByName(c.Children[i].Key); field.IsValid() && field.CanInterface() {
+				fieldPtr := field.Addr().Interface()
+
+				if err := c.Children[i].Unmarshal(fieldPtr); err != nil {
+					return fmt.Errorf("while unmarshalling child config with key %s: %s", c.Children[i].Key, err)
+				}
+
+			}
+		}
+		return nil
+	case reflect.Slice:
+		switch rv.Elem().Kind() {
+		case reflect.Struct:
+			// Create a temporary Value of the same type as dereferenced value, then get a Value of the same type as
+			// its elements. Unmarshal into that Value and append the temporary Value to the original.
+			tv := reflect.New(drv.Type()).Elem()
+			if err := storeStructConfigValues(c.Values, tv); err != nil {
+				return fmt.Errorf("while unmarshalling values into %s: %s", drv.Type(), err)
+			}
+			for i := range c.Children {
+				// If a config has children but the struct type that is an element of this slice has no corresponding
+				// field the child is dropped without notice.
+				if field := tv.FieldByName(c.Children[i].Key); field.IsValid() {
+					if err := c.Children[i].Unmarshal(field); err != nil {
+						return fmt.Errorf("while unmarshalling child config with key %s: %s", c.Children[i].Key, err)
+					}
+				}
+			}
+			drv.Set(reflect.Append(drv, tv))
+			return nil
+		default:
+			for i := range c.Values {
+				tv := reflect.New(drv.Type().Elem()).Elem()
+				if err := storeConfigValue(c.Values[i], tv); err != nil {
+					return fmt.Errorf("while unmarhalling values into %s: %s", drv.Type(), err)
+				}
+				drv.Set(reflect.Append(drv, tv))
+			}
+			return nil
+		}
+	default: // Kind is one of the number, bool or string kinds
+		if len(c.Values) != 1 {
+			return fmt.Errorf("cannot unmarshal config with %d values into scalar type %s", len(c.Values), drv.Type())
+		}
+		return storeConfigValue(c.Values[0], drv)
+	}
+}
+
+func storeConfigValue(cv ConfigValue, v reflect.Value) error {
+	rvt := v.Type()
+	var cvt reflect.Type
+	var cvv reflect.Value
+
+	switch cv.typ {
+	case configTypeString:
+		cvt = reflect.TypeOf(cv.s)
+		cvv = reflect.ValueOf(cv.s)
+	case configTypeBoolean:
+		cvt = reflect.TypeOf(cv.b)
+		cvv = reflect.ValueOf(cv.b)
+	case configTypeNumber:
+		cvt = reflect.TypeOf(cv.f)
+		cvv = reflect.ValueOf(cv.f)
+	default:
+		panic("received ConfigValue with unknown type")
+	}
+
+	if cvt.ConvertibleTo(rvt) {
+		v.Set(cvv.Convert(rvt))
+		return nil
+	}
+	if v.Kind() == reflect.Slice && cvt.ConvertibleTo(rvt.Elem()) {
+		v.Set(reflect.Append(v, cvv.Convert(rvt.Elem())))
+		return nil
+	}
+	return fmt.Errorf("cannot unmarshal %s type config value to type %s", cv.typ, v.Type())
+}
+
+func storeStructConfigValues(cv []ConfigValue, v reflect.Value) error {
+	args := v.FieldByName("Args")
```

Comment:
It is legal (and common) for a block ("Config") to have no values. Before checking if an "Args" field exists, add a short-path for `len(cv) == 0`.